src/test/java/com/openshift/internal/restclient/okhttp/ClientProxyConnectionTest.java

```diff
@@ -0,0 +1,120 @@
+
+/**
+ * @author Vlad Slepukhin
+ */
+public class ClientProxyConnectionTest {
+
+	private static final Logger LOG = LoggerFactory.getLogger(ClientProxyConnectionTest.class);
+
+
+	private static final String OPENSHIFT_URL = "https://openshift.test.com";
+	private static final String PROXY_HOST = "127.0.0.1";
+	private static final int PROXY_PORT = 8080;
+
+	private class TestProxySelector extends ProxySelector {
+
+		@Override
+		public List<Proxy> select(URI uri) {
+			List<Proxy> result = new ArrayList<>();
+			result.add(new Proxy(Proxy.Type.HTTP, new InetSocketAddress(PROXY_HOST, PROXY_PORT)));
+			return result;
+		}
+
+		@Override
+		public void connectFailed(URI uri, SocketAddress sa, IOException ioe) {
+		}
+	}
+
+	@Test
+	public void testProxySetInOkHttp() throws NoSuchFieldException, IllegalAccessException {
+		Proxy expectedProxy = new Proxy(Proxy.Type.HTTP, new InetSocketAddress(PROXY_HOST, PROXY_PORT));
+		IClient client = new ClientBuilder(OPENSHIFT_URL)
+				.proxy(expectedProxy)
+				.build();
+
+		OkHttpClient okClient = getOkHttpClient(client);
+		Proxy proxyInOkClient = okClient.proxy();
+
+		LOG.info(proxyInOkClient.toString());
+
+		assertEquals(expectedProxy.type(), proxyInOkClient.type());
+		assertEquals(expectedProxy.address(), proxyInOkClient.address());
+	}
+
+	@Test
+	public void testProxyNotSetInOkHttp() throws NoSuchFieldException, IllegalAccessException {
+		IClient client = new ClientBuilder(OPENSHIFT_URL)
+				.build();
+
+		OkHttpClient okClient = getOkHttpClient(client);
+		Proxy proxyInOkClient = okClient.proxy();
+
+		assertNull(proxyInOkClient);
+	}
+
+	@Test
+	public void testProxySelectorSetInOkHttp() throws NoSuchFieldException, IllegalAccessException {
+		ProxySelector expectedSelector = new TestProxySelector();
+		IClient client = new ClientBuilder(OPENSHIFT_URL)
+				.proxySelector(expectedSelector)
+				.build();
+
+
+		OkHttpClient okClient = getOkHttpClient(client);
+		ProxySelector proxySelector = okClient.proxySelector();
+
+		assertNotSame(proxySelector, ProxySelector.getDefault());
+		assertEquals(expectedSelector, proxySelector);
+
+		Proxy expectedProxyFromList = new Proxy(Proxy.Type.HTTP, new InetSocketAddress(PROXY_HOST, PROXY_PORT));
+		Proxy actualProxyFromList = proxySelector.select(null).get(0);
+		assertEquals(expectedProxyFromList.type(), actualProxyFromList.type());
+		assertEquals(expectedProxyFromList.address(), actualProxyFromList.address());
+	}
+
+	@Test
+	public void testProxySelectorNotSetInOkHttp() throws NoSuchFieldException, IllegalAccessException {
+		IClient client = new ClientBuilder(OPENSHIFT_URL)
+				.build();
+
+		OkHttpClient okClient = getOkHttpClient(client);
+		ProxySelector proxySelector = okClient.proxySelector();
+
+		assertEquals(proxySelector, ProxySelector.getDefault());
+	}
+
+
+	private OkHttpClient getOkHttpClient(IClient client) throws NoSuchFieldException, IllegalAccessException {
+		Field clientField = client.getClass().getDeclaredField("client");
```

Comment:
Any reason to not cast this to the impl of IClient, return the value directly instead of using reflection?  What advantage here is there in using reflection?

Reply:
Third option is to make OkHttpClient protected. WDYT?

Reply:
@jcantrill, is it better now?

Reply:
yes